File: backend/app/models/text_analysis.py
from pydantic import BaseModel, Field
from typing import List, Optional, Dict

class TextStats(BaseModel):
    word_count: int
    sentence_count: int
    avg_word_length: float
    avg_sentence_length: float
    readability_score: float

class ToneAnalysis(BaseModel):
    primary_tone: str
    tone_scores: Dict[str, float]

class GrammarIssue(BaseModel):
    type: str
    text: str
    correction: str
    explanation: str

class StyleSuggestion(BaseModel):
    aspect: str
    current: str
    improvement: str
    rationale: str

class StructureIssue(BaseModel):
    type: str
    location: str
    suggestion: str
    rationale: str

class Analysis(BaseModel):
    grammar: Optional[Dict] = None
    style: Optional[Dict] = None
    structure: Optional[Dict] = None

class TextAnalysisRequest(BaseModel):
    text: str
    style: str = Field(default="professional", description="The desired writing style")
    focus_areas: List[str] = Field(
        default=["grammar", "style", "structure"],
        description="Areas to focus on during analysis"
    )

class TextAnalysisResponse(BaseModel):
    original_text: str
    improved_text: str
    text_stats: TextStats
    tone_analysis: ToneAnalysis
    analysis: Analysis
    improvements: List[str] = [] 

File: types/text-analysis.ts
export interface GrammarIssue {
  type: string;
  text: string;
  explanation: string;
  position?: {
    start: number;
    end: number;
  };
}

export interface StyleSuggestion {
  aspect: string;
  current: string;
  improvement: string;
  rationale?: string;
}

export interface StructureIssue {
  location: string;
  suggestion: string;
}

export interface StyleFeedback {
  tone: string;
  readability_score: number;
  complexity: string;
  style_suggestions: StyleSuggestion[];
}

export interface ProactiveFeedback {
  type: string;
  description: string;
  suggestion: string;
  area: string;
}

export interface GoalAchievement {
  achieved: boolean;
  score: number;
  remaining_steps: string[];
}

export interface Suggestion {
  type: string;
  description: string;
  pros: string[];
  cons: string[];
  implementation: string;
}

export interface TextStats {
  word_count: number;
  sentence_count: number;
  avg_word_length: number;
  avg_sentence_length: number;
  readability_score: number;
}

export interface ToneAnalysis {
  primary_tone: string;
  tone_scores: {
    [key: string]: number;
  };
}

export interface AnalysisResult {
  error?: string;
  grammar_issues: GrammarIssue[];
  suggestions: Suggestion[];
  style_feedback: StyleFeedback;
  improved_text: string;
  proactive_feedback: ProactiveFeedback[];
  goal_achievement: GoalAchievement;
  text_stats: TextStats;
  tone_analysis: ToneAnalysis;
  analysis?: {
    grammar?: {
      confidence_score: number;
      issues: GrammarIssue[];
    };
    style?: {
      style_score: number;
      suggestions: StyleSuggestion[];
    };
    structure?: {
      structure_score: number;
      flow_issues: StructureIssue[];
    };
  };
}

export type OutputStyle = "grammar" | "friendly" | "professional" | "concise";
